visualization/app/codeCharta/state/store/metricData/nodeMetricData/nodeMetricData.service.ts

```diff
@@ -1,77 +1,69 @@
-import { isActionOfType } from "../../../../util/reduxHelper"
-import { BlacklistItem, NodeMetricData } from "../../../../codeCharta.model"
+import { NodeMetricData } from "../../../../codeCharta.model"
 import { FilesSelectionSubscriber, FilesService } from "../../files/files.service"
 import { BlacklistService, BlacklistSubscriber } from "../../fileSettings/blacklist/blacklist.service"
 import { AttributeTypesService, AttributeTypesSubscriber } from "../../fileSettings/attributeTypes/attributeTypes.service"
-import { FileState } from "../../../../model/files/files"
+import { nodeMetricDataSelector } from "../../../selectors/accumulatedData/metricData/nodeMetricData.selector"
 
 export interface NodeMetricDataSubscriber {
 	onNodeMetricDataChanged(nodeMetricData: NodeMetricData[])
 }
 
-export class NodeMetricDataService implements StoreSubscriber, FilesSelectionSubscriber, BlacklistSubscriber, AttributeTypesSubscriber {
+export class NodeMetricDataService implements FilesSelectionSubscriber, BlacklistSubscriber, AttributeTypesSubscriber {
 	static UNARY_METRIC = "unary"
 	private static NODE_METRIC_DATA_CHANGED_EVENT = "node-metric-data-changed"
 
+	private nodeMetricData: NodeMetricData[]
+
 	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
 		"ngInject"
-		StoreService.subscribe(this.$rootScope, this)
 		FilesService.subscribe(this.$rootScope, this)
 		BlacklistService.subscribe(this.$rootScope, this)
 		AttributeTypesService.subscribe(this.$rootScope, this)
 	}
 
-	onStoreChanged(actionType: string) {
-		if (isActionOfType(actionType, NodeMetricDataActions)) {
-			this.notify(this.select())
-		}
-	}
-
-	onFilesSelectionChanged(files: FileState[]) {
-		this.storeService.dispatch(calculateNewNodeMetricData(files, this.storeService.getState().fileSettings.blacklist))
+	onFilesSelectionChanged() {
+		this.updateNodeMetricData()
 	}
 
-	onBlacklistChanged(blacklist: BlacklistItem[]) {
-		this.storeService.dispatch(calculateNewNodeMetricData(this.storeService.getState().files, blacklist))
+	onBlacklistChanged() {
+		this.updateNodeMetricData()
 	}
 
+	// This shouldn't be needed, as nodeMetricData is not dependent on AttributeTypesChanged, but switching to median in attribute side bar breaks without it
 	onAttributeTypesChanged() {
-		this.storeService.dispatch(
-			calculateNewNodeMetricData(this.storeService.getState().files, this.storeService.getState().fileSettings.blacklist)
-		)
+		this.$rootScope.$broadcast(NodeMetricDataService.NODE_METRIC_DATA_CHANGED_EVENT, { nodeMetricData: this.nodeMetricData })
 	}
 
 	getMetrics() {
-		return this.storeService.getState().metricData.nodeMetricData.map(x => x.name)
+		return this.nodeMetricData.map(x => x.name)
 	}
 
 	isMetricAvailable(metricName: string) {
-		return this.storeService.getState().metricData.nodeMetricData.some(x => x.name === metricName)
+		return this.nodeMetricData?.some(x => x.name === metricName)
```

Comment:
This changes the return type in case `nodeMetricData` has not yet been set from `false` to `undefined`. I suggest to just use an empty array as default value. That way the return type is not changed and there's no need for the optional chaining anymore.